### PR TITLE
stats: resolve the constant selectivity when handling DNF exprs

### LINKED ISSUE #31096: Make `Selectivity()` correctly handle `true` and `false`
## Enhancement
```sql
create table t1(a int, b int, c int);
insert into t1 value(10,10,10);
insert into t1 select * from t1;
insert into t1 select * from t1;
insert into t1 select * from t1;
insert into t1 select * from t1;
insert into t1 select * from t1;
insert into t1 select * from t1;
insert into t1 select * from t1;
insert into t1 value(1,1,1);
analyze table t1;
```

The row count estimation of the following two SQL should be the same:

```
> explain select * from t1 where a=1 or b=1;
+-------------------------+---------+-----------+---------------+------------------------------------------+
| id                      | estRows | task      | access object | operator info                            |
+-------------------------+---------+-----------+---------------+------------------------------------------+
| TableReader_7           | 0.26    | root      |               | data:Selection_6                         |
| └─Selection_6           | 0.26    | cop[tikv] |               | or(eq(test2.t1.a, 1), eq(test2.t1.b, 1)) |
|   └─TableFullScan_5     | 129.00  | cop[tikv] | table:t1      | keep order:false                         |
+-------------------------+---------+-----------+---------------+------------------------------------------+



> explain select * from t1 where 0=1 or a=1 or b=1;
+-------------------------+---------+-----------+---------------+-------------------------------------------------+
| id                      | estRows | task      | access object | operator info                                   |
+-------------------------+---------+-----------+---------------+-------------------------------------------------+
| TableReader_7           | 103.25  | root      |               | data:Selection_6                                |
| └─Selection_6           | 103.25  | cop[tikv] |               | or(0, or(eq(test2.t1.a, 1), eq(test2.t1.b, 1))) |
|   └─TableFullScan_5     | 129.00  | cop[tikv] | table:t1      | keep order:false                                |
+-------------------------+---------+-----------+---------------+-------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: ailinkid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close #31096 

Problem Summary:

### What is changed and how it works?
For constant expr among a DNF, we should regard it's selectivity as 1 or 0 when considering DNF selectivity
- explain select * from t1 where 0=1 or a=1 or b=1;
- explain select * from t1 where false or a=1 or b=1;
- explain select * from t1 where "false" or a=1 or b=1;
- explain select * from t1 where 0 or a=1 or b=1;
- explain select * from t1 where null or a=1 or b=1;

should also be the same as `explain select * from t1 where a=1 or b=1;`

Additional, we treat `explain select * from t1 where 1 or a=1 or b=1` as selectivity = 1

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -5195,3 +5195,32 @@ func (s *testIntegrationSuite) TestIssue20510(c *C) {
...
+	}
+	testKit.MustExec("insert into t1 value(1,1,1)")
+	testKit.MustExec("analyze table t1")
+
+	testKit.MustQuery("explain select * from t1 where a=1 or b=1;").Check(testkit.Rows("TableReader_7 1.99 root  data:Selection_6]\n" +
+		"[└─Selection_6 1.99 cop[tikv]  or(eq(test.t1.a, 1), eq(test.t1.b, 1))]\n" +
+		"[  └─TableFullScan_5 129.00 cop[tikv] table:t1 keep order:false"))
+	testKit.MustQuery("explain select * from t1 where 0=1 or a=1 or b=1;").Check(testkit.Rows("TableReader_7 1.99 root  data:Selection_6]\n" +
+		"[└─Selection_6 1.99 cop[tikv]  or(0, or(eq(test.t1.a, 1), eq(test.t1.b, 1)))]\n" +
+		"[  └─TableFullScan_5 129.00 cop[tikv] table:t1 keep order:false"))
```
Comment (xuyifangreeneyes) [inline planner/core/integration_test.go:5218]: Could you add more cases such as `false`, `"false"`, 'null'?

File: statistics/selectivity.go [hunk 2]
```diff
@@ -354,6 +354,20 @@ func (coll *HistColl) Selectivity(ctx sessionctx.Context, exprs []expression.Exp
 				if ok {
 					continue
 				}
+				// where {"0" / 0 / "false" / false / null} or A or B ... the '0' constant item should be ignored.
+				if c, ok := cond.(*expression.Constant); ok {
```
Comment (time-and-fate) [inline statistics/selectivity.go:358]: According to our logic in `buildSelection` and `DeleteTrueExprs`, I think we should (1) avoid `DeferredExpr` and `ParamMarker` and (2) use `EvalBool()` or `ToBool()` to convert it to a bool.

Comment (AilinKid) [reply statistics/selectivity.go:358]: nice suggestion, updated

File: statistics/selectivity.go [hunk 3]
```diff
@@ -354,6 +354,24 @@ func (coll *HistColl) Selectivity(ctx sessionctx.Context, exprs []expression.Exp
...
+					if !expression.MaybeOverOptimized4PlanCache(ctx, []expression.Expression{cond}) {
+						if c.Value.IsNull() {
+							// constant == 0
```
Comment (time-and-fate) [inline statistics/selectivity.go:361]:
```suggestion
							// constant is null
```

**Review by time-and-fate — APPROVED** [review]

**Review by xuyifangreeneyes — APPROVED** [review]

**Pull request merged**